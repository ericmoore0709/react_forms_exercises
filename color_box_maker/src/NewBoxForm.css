.BoxForm {
    align-items: center;
    background-color: beige;
    display: flex;
    flex-direction: column;
    margin: auto;
    padding: 0.5em;
    width: max-content;
}

.BoxForm > button {
    background-color: blue;
    border: 1px solid black;
    border-radius: .5rem;
    color: white;
}

.input-container {
    display: flex;
    justify-content: space-between;
    margin: auto;
}