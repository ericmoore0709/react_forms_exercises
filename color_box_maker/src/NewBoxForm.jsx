import { useState } from 'react';
import './NewBoxForm.css';

const NewBoxForm = () => {

    const INITIAL_DATA = {
        color: '#ffffff',
        width: 50,
        height: 50
    }

    const [formData, setFormData] = useState(INITIAL_DATA);

    const handleChange = (evt) => {

        const { name, value } = evt.target;

        setFormData({
            ...formData,
            [name]: value
        });
    }

    const handleSubmit = (e) => {
        e.preventDefault();
    }

    return (
        <form className="BoxForm" onSubmit={handleSubmit}>
            <div className="input-container">
                <label htmlFor="color">Color</label>
                <input
                    id="color"
                    name="color"
                    type="color"
                    placeholder="Color"
                    value={formData.color}
                    onChange={handleChange}
                />
            </div>
            <div className="input-container">
                <label htmlFor="width">Width</label>
                <input
                    id="width"
                    name="width"
                    type="number"
                    min={10}
                    max={300}
                    placeholder="Width"
                    value={formData.width}
                    onChange={handleChange}
                />
            </div>
            <div className="input-container">
                <label htmlFor="height">Height</label>
                <input
                    id="height"
                    name="height"
                    type="number"
                    min={10}
                    max={300}
                    placeholder="Height"
                    value={formData.height}
                    onChange={handleChange}
                />
            </div>
            <button>Add Box</button>
        </form>
    )
}

export default NewBoxForm;