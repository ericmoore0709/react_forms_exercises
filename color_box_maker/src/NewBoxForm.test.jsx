import { expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';

it('should not smoke', () => {
    render(<NewBoxForm addBox={vi.fn()} />);
});

it('should match snapshot', () => {
    const { asFragment } = render(<NewBoxForm addBox={vi.fn()} />);
    expect(asFragment()).toMatchSnapshot();
});

it('should fire addBox on submit', () => {
    const addMock = vi.fn();
    render(<NewBoxForm addBox={addMock} />);

    fireEvent.change(screen.getByLabelText('Color'), { target: { value: '#ff0000' } });
    fireEvent.change(screen.getByLabelText('Width'), { target: { value: '100' } });
    fireEvent.change(screen.getByLabelText('Height'), { target: { value: '200' } });

    fireEvent.click(screen.getByRole('button', { name: /add box/i }));
    expect(addMock).toBeCalledWith({ color: '#ff0000', width: '100', height: '200' });
});

it('should change color when color input is changed', () => {
    render(<NewBoxForm addBox={vi.fn()} />);
    const colorInput = screen.getByLabelText('Color');

    fireEvent.change(colorInput, { target: { value: '#ff0000' } });
    expect(colorInput.value).toBe('#ff0000');
});

it('should change width when width input is changed', () => {
    render(<NewBoxForm addBox={vi.fn()} />);
    const widthInput = screen.getByLabelText('Width');

    fireEvent.change(widthInput, { target: { value: '200' } });
    expect(widthInput.value).toBe('200');
});

it('should change height when height input is changed', () => {
    render(<NewBoxForm addBox={vi.fn()} />);
    const heightInput = screen.getByLabelText('Height');

    fireEvent.change(heightInput, { target: { value: '70' } });
    expect(heightInput.value).toBe('70');
});

it('should reset form after submit', () => {
    const addMock = vi.fn();
    render(<NewBoxForm addBox={addMock} />);

    fireEvent.change(screen.getByLabelText('Color'), { target: { value: '#ff0000' } });
    fireEvent.change(screen.getByLabelText('Width'), { target: { value: '100' } });
    fireEvent.change(screen.getByLabelText('Height'), { target: { value: '200' } });

    fireEvent.click(screen.getByRole('button', { name: /add box/i }));

    expect(screen.getByLabelText('Color').value).toBe('#ffffff');
    expect(screen.getByLabelText('Width').value).toBe('50');
    expect(screen.getByLabelText('Height').value).toBe('50');
});
