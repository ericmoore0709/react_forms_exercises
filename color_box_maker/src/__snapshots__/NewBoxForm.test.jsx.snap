// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="BoxForm"
  >
    <div
      class="input-container"
    >
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        placeholder="Color"
        type="color"
        value="#ffffff"
      />
    </div>
    <div
      class="input-container"
    >
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        max="300"
        min="10"
        name="width"
        placeholder="Width"
        type="number"
        value="50"
      />
    </div>
    <div
      class="input-container"
    >
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        max="300"
        min="10"
        name="height"
        placeholder="Height"
        type="number"
        value="50"
      />
    </div>
    <button>
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
